File: core/src/core2/sql/parser.clj
(ns core2.sql.parser
  (:require [clojure.java.io :as io]
            [clojure.set :as set]
            [clojure.string :as str]
            [core2.error :as err]
            [core2.util :as util])
  (:import [core2.sql.parser Parser Parser$AParser Parser$AltParser Parser$CatParser Parser$EpsilonParser Parser$HideParser Parser$MemoTable Parser$MemoizeLeftRecParser Parser$MemoizeParser Parser$NegParser Parser$NonTerminalParser Parser$OptParser Parser$OrdParser Parser$ParseErrors Parser$RegexpParser Parser$RepeatParser Parser$RuleParser Parser$StringParser]
           [java.util ArrayDeque HashSet Map Set]
           java.util.function.Function
           [java.util.regex Matcher Pattern]))

;; https://arxiv.org/pdf/1509.02439v1.pdf
;; https://medium.com/@gvanrossum_83706/left-recursive-peg-grammars-65dab3c580e1
;; https://bford.info/pub/lang/thesis.pdf
;; https://arxiv.org/pdf/1806.11150.pdf

(set! *unchecked-math* :warn-on-boxed)

(defrecord ParseFailure [^String in ^Set errs ^int idx])

(defn- left-recursive? [grammar rule-name]
  (let [visited (HashSet.)
        stack (ArrayDeque.)]
    (.push stack (get grammar rule-name))
    (loop []
      (if-let [node (.poll stack)]
        (cond
          (= (:keyword node) rule-name)
          true

          (.contains visited node)
          (recur)

          :else
          (do (.add visited node)
              (let [node (if (= :nt (:tag node))
                           (get grammar (:keyword node))
                           node)]
                (case (:tag node)
                  :cat (let [[opts mandatory] (split-with (comp #{:opt :star} :tag) (:parsers node))]
                         (doseq [opt opts]
                           (.push stack opt))
                         (when-not (empty? mandatory)
                           (.push stack (first mandatory))))
                  :ord (do (.push stack (:parser1 node))
                           (.push stack (:parser2 node)))
                  :alt (doseq [alt (:parsers node)]
                         (.push stack alt))
                  (:star :plus :opt) (.push stack (:parser node))
                  :nt (.push stack node)
                  nil)
                (recur))))
        false))))

(defn- rule-kw->name [kw]
  (str "<" (str/replace (name kw) "_" " ") ">"))

(defn index->line-column [^String in ^long idx]
  (loop [line 1
         col 1
         n 0]
    (cond
      (= idx n) {:line line :column col}
      (= \newline (.charAt in n)) (recur (inc line) 1 (inc n))
      :else (recur line (inc col) (inc n)))))

(defn failure? [x]
  (instance? ParseFailure x))

(defn failure->str ^String [^ParseFailure failure]
  (let [{:keys [^long line ^long column]} (index->line-column (.in failure) (.idx failure))]
    (with-out-str
      (println (str "Parse error at line " line ", column " column ":"))
      (println (get (str/split-lines (.in failure)) (dec line)))
      (dotimes [_ (dec column)]
        (print " "))
      (println "^")
      (doseq [[category errs] (group-by first (.errs failure))]
        (println (str (str/capitalize (name category))
                      (if (= 1 (count errs))
                        ":"
                        " one of:")))
        (doseq [[_ msg] errs]
          (println msg))))))

(defn- surround-parser-with-ws ^core2.sql.parser.Parser$AParser [parser ^Pattern ws-pattern]
  (Parser$CatParser. [(Parser$HideParser. (Parser$StringParser. "" ws-pattern))
                      parser
                      (Parser$EpsilonParser.)]))

(defn- lookahead [grammar seen {:keys [tag] :as parser}]
  (when-not (contains? seen parser)
    (let [seen (conj seen parser)]
      (case tag
        :nt (recur grammar seen (get grammar (:keyword parser)))
        (:star :opt :neg :epsilon) #{::undecided-lookahead}
        :plus (recur grammar seen (:parser parser))
        :cat ((fn step [[parser & parsers]]
                (case (:tag parser)
                  (:star :opt) (set/union (lookahead grammar seen (:parser parser))
                                          (step parsers))
                  :neg (step parsers)
                  (when parser
                    (lookahead grammar seen parser))))
              (:parsers parser))
        :alt (reduce set/union (map (partial lookahead grammar seen) (:parsers parser)))
        :ord (let [parsers ((fn step [{:keys [parser1 parser2]}]
                              (cons parser1 (if (= :ord (:tag parser2))
                                              (step parser2)
                                              [parser2])))
                            parser)]
               (reduce set/union (map (partial lookahead grammar seen) parsers)))
        :regexp (if-let [lookahead-set (not-empty (set (for [n (range 128)
                                                             :let [m (re-matcher (:regexp parser) (str (char n)))]
                                                             :when (or (.lookingAt m)
                                                                       (.hitEnd m))]
                                                         (char n))))]
                  lookahead-set
                  #{::undecided-lookahead})
        :string (hash-set (Character/toLowerCase (char (first (:string parser))))
                          (Character/toUpperCase (char (first (:string parser)))))))))

(def ^:private undecided-lookahead-table (boolean-array 128 true))

(defn- lookahead-alt-table [grammar parsers]
  (vec (for [parser parsers
             :let [lookahead-set (lookahead grammar #{} parser)]]
         (if (contains? lookahead-set ::undecided-lookahead)
           undecided-lookahead-table
           (reduce
            (fn [^booleans acc ^long c]
              (if (< c 128)
                (doto acc
                  (aset c true))
                acc))
            (boolean-array 128)
            (map long lookahead-set))))))

(defn build-ebnf-parser
  ([grammar ws-pattern]
   (build-ebnf-parser grammar ws-pattern (fn [_rule-name]
                                           Parser/NEVER_RAW)))
  ([^Map grammar ws-pattern ->raw?]
   (let [^Map rule->id (zipmap (keys grammar) (range))
         ^"[Lcore2.sql.parser.Parser$AParser;" rules (make-array Parser$AParser (count rule->id))]
     (letfn [(build-parser [rule-name {:keys [tag hide] :as parser}]
               (let [parser (case tag
                              :nt (let [body (get grammar (:keyword parser))]
                                    (if (and (or hide (= {:reduction-type :raw} (:red body)))
                                             (contains? #{:string :regexp :nt} (:tag body)))
                                      (build-parser rule-name body)
                                      (Parser$NonTerminalParser. (get rule->id (:keyword parser)))))
                              :star (Parser$RepeatParser. (build-parser rule-name (:parser parser)) true)
                              :plus (Parser$RepeatParser. (build-parser rule-name (:parser parser)) false)
                              :opt (Parser$OptParser. (build-parser rule-name (:parser parser)))
                              :cat (Parser$CatParser. (mapv (partial build-parser rule-name) (:parsers parser)))
                              :alt (Parser$AltParser. (mapv (partial build-parser rule-name) (:parsers parser))
                                                      (lookahead-alt-table grammar (:parsers parser)))
                              :ord (let [parsers ((fn step [{:keys [parser1 parser2]}]
                                                    (cons parser1 (if (= :ord (:tag parser2))
                                                                    (step parser2)
                                                                    [parser2])))
                                                  parser)]
                                     (Parser$OrdParser. (mapv (partial build-parser rule-name) parsers)))
                              :neg (Parser$NegParser. (build-parser rule-name (:parser parser)))
                              :epsilon (Parser$EpsilonParser.)
                              :regexp (Parser$RegexpParser. (:regexp parser)
                                                            [:expected (rule-kw->name rule-name)]
                                                            (reify Function
                                                              (apply [_ m]
                                                                [(.group ^Matcher m)]))
                                                            ws-pattern)
                              :string (if (re-find #"^\w+$" (:string parser))
                                        (Parser$RegexpParser. (Pattern/compile (str (Pattern/quote (:string parser)) "\\b")
                                                                               Pattern/CASE_INSENSITIVE)
                                                              [:expected (:string parser)]
                                                              (let [ast [(:string parser)]]
                                                                (reify Function
                                                                  (apply [_ m]
                                                                    ast)))
                                                              ws-pattern)
                                        (Parser$StringParser. (:string parser) ws-pattern)))]
                 (if hide
                   (Parser$HideParser. parser)
                   parser)))]
       (doseq [[k v] grammar
               :let [rule-id (int (get rule->id k))
                     raw? (if (= {:reduction-type :raw} (:red v))
                            Parser/ALWAYS_RAW
                            (->raw? k))
                     parser (Parser$RuleParser. k raw? (build-parser k v))
                     parser (if (left-recursive? grammar k)
                              (Parser$MemoizeLeftRecParser. parser rule-id (count grammar))
                              (Parser$MemoizeParser. parser rule-id (count grammar)))]]
         (aset rules rule-id ^Parser$AParser parser))

       (doseq [^Parser$AParser rule-parser rules]
         (.init rule-parser rules))

       (fn [^String in start-rule]
         (let [memos (Parser$MemoTable. (.length in))
               parser (surround-parser-with-ws (aget rules (.get rule->id start-rule)) ws-pattern)]
           (if-let [state (.parse parser in 0 memos Parser/NULL_PARSE_ERRORS false)]
             (.nth (.ast state) 0 nil)
             (let [memos (Parser$MemoTable. (.length in))
                   errors (Parser$ParseErrors.)]
               (.parse parser in 0 memos errors true)
               (ParseFailure. in (.getErrors errors) (.getIndex errors))))))))))

(defn or-throw [{:keys [errs] :as res}]
  (if errs
    (let [failure-str (failure->str res)]
      (throw (err/illegal-arg :core2.sql/parse-error
                              {::err/message (str "Invalid SQL query: " failure-str)
                               :errs [failure-str]})))
    res))

(def sql-cfg
  (read-string (slurp (io/resource "core2/sql/parser/SQL2011.edn"))))

(def sql-regular-ws-pattern
  #"\s+")

(def sql-single-line-comment-pattern
  #"--.*")

(def sql-multi-line-comment-pattern
  #"(?s)/[*].*?(?:[*]/|$)(?-s)")

;; TODO: I think this should be plus and not star, as the idea is that
;; it should match mandatory whitespace, but detect boundaries and
;; punctuation as well. Most of that is really handled by the parsing
;; of the keywords themselves, enforcing them to be at word
;; boundaries. Identifiers may bleed over into keywords if there's no
;; boundary, like fooASbar is a single identifier.

;; NOTE, the following are currently not used.

(def sql-after-punctuation-pattern
  #"(?<=\p{Punct})")

(def sql-word-boundary-pattern
  #"\b")


(def sql-ws-pattern
  (Pattern/compile
   (str "(?:"
        (->> (for [^Pattern p [sql-regular-ws-pattern
                               sql-single-line-comment-pattern
                               sql-multi-line-comment-pattern]]
               (.pattern p))
             (str/join "|"))
        ")*")))

;; API

(def sql-parser
  (build-ebnf-parser sql-cfg
                     sql-ws-pattern
                     (fn [rule-name]
                       (if (and (not (contains? #{:table_primary :query_expression :table_expression} rule-name))
                                (re-find #"(^|_)(term|factor|primary|expression|query_expression_body|boolean_test)$" (name rule-name)))
                         Parser/SINGLE_CHILD
                         Parser/NEVER_RAW))))

(def parse
  (-> (fn self
        ([in] (self in :directly_executable_statement))
        ([in start-rule]
         (-> (sql-parser in start-rule)
             (vary-meta assoc :sql in))))
      (util/lru-memoize)))

(comment
  (sql-parser "SELECT foo.bar,
                        SUM(foo.bar) OVER (PARTITION BY foo.biff) AS baz
                        FROM foo" :directly_executable_statement)


  (sql-parser "SELECT * FROMfoo" :directly_executable_statement)

  (println (failure->str (sql-parser "* FROMfoo" :term)))

  (sql-parser "  (SELECT avg(c) FROM t1)  " :subquery)

  (sql-parser "a[0]" :value_expression_primary)

  (sql-parser "- - 2" :factor)

  (time (sql-parser "(((((((((1)))))))))" :value_expression_primary))

  (= (sql-parser "SELECT pk FROM tab0 WHERE ((col0 >= 6) OR (col1 = 2.11 AND col3 BETWEEN 3 AND 4) AND ((col1 > 9.58 AND col0 <= 4)) AND (col0 >= 6))" :directly_executable_statement)
     (sql-parser (core2.sql.logic-test.xtdb-engine/remove-unnecessary-parens "SELECT pk FROM tab0 WHERE ((col0 >= 6) OR (col1 = 2.11 AND col3 BETWEEN 3 AND 4) AND ((col1 > 9.58 AND col0 <= 4)) AND (col0 >= 6))") :directly_executable_statement))

  (sql-parser "IN ( col0 * col0 )" :in_predicate_part_2)

  (sql-parser "INTERVAL '3 4' DAY TO HOUR" :interval_literal)

  (doseq [sql ["SELECT u.a[0] AS first_el FROM uo"
               "SELECT u.b[u.a[0]] AS dyn_idx FROM u"
               "SELECT 1 YEAR + 3 MONTH + 4 DAY t FROM foo WHERE foo.a = 42"
               "SELECT 3 * 1 YEAR t FROM foo WHERE foo.a = 42"
               "SELECT foo.a YEAR + 1 MONTH + 2 DAY t FROM foo WHERE foo.a = 42"
               "SELECT foo.a YEAR + 1 MONTH - 2 DAY t FROM foo WHERE foo.a = 42"
               "SELECT foo.a || 'a' || 'b' t FROM foo WHERE foo.a = 42"
               "SELECT ALL 4 AS col1 FROM tab2 AS cor0 WHERE NULL IS NULL"
               "SELECT ALL 74 * - COALESCE ( + CASE - CASE WHEN NOT ( NOT - 79 >= NULL ) THEN 48 END WHEN + + COUNT ( * ) THEN 6 END, MIN ( ALL + - 30 ) * 45 * 77 ) * - 14 FROM (VALUES 0) AS no_from"
               "SELECT cor0.col2 AS col2 FROM tab2 AS cor0 GROUP BY col2 HAVING NOT NULL < NULL"
               "SELECT DISTINCT col2 FROM tab1 WHERE NULL BETWEEN NULL AND - col2"
               ;; spec doesn't seem to allow expressions inside IN.
               "SELECT * FROM tab1 WHERE NULL NOT IN ( col0 * col0 )"
               "SELECT INTERVAL '3 4' DAY TO HOUR t FROM foo WHERE foo.a = 42"
               "SELECT * FROMfoo"
               "SELECT + NULLIF ( + 67, + + NULLIF ( + + 46, 66 - CASE WHEN 51 IN ( + 91 ) THEN + SUM ( CAST ( NULL AS INTEGER ) ) + - 92 ELSE - ( - 47 ) END ) ) FROM (VALUES 0) AS no_from
"]
          :let [tree (sql-parser sql :directly_executable_statement)]
          :when (failure? tree)]
    (println (failure->str tree)))

  (let [in "SELECT CASE WHEN c>(SELECT avg(c) FROM t1) THEN a*2 ELSE b*10 END,
       a+b*2+c*3+d*4,
       abs(b-c),
       (SELECT count(*) FROM t1 AS x WHERE x.b<t1.b),
       c-d,
       a+b*2,
       CASE WHEN a<b-3 THEN 111 WHEN a<=b THEN 222
        WHEN a<b+3 THEN 333 ELSE 444 END
  FROM t1 WHERE e+d BETWEEN a+b-10 AND c+130
    OR coalesce(a,b,c,d,e)<>0"]

    (time
     (dotimes [_ 1000]
       (sql-parser in :directly_executable_statement)))))

File: core/src/core2/sql/parser/Parser.java
package core2.sql.parser;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.HashSet;
import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.util.function.Function;
import java.util.function.Predicate;
import clojure.lang.IObj;
import clojure.lang.IPersistentMap;
import clojure.lang.IPersistentVector;
import clojure.lang.ITransientCollection;
import clojure.lang.Keyword;
import clojure.lang.PersistentArrayMap;
import clojure.lang.PersistentVector;

public final class Parser {

    public static final class ParseState {
        public final IPersistentVector ast;
        public final int idx;

        public ParseState(final IPersistentVector ast, final int idx) {
            this.ast = ast;
            this.idx = idx;
        }
    }

    public interface IParseErrors {
        void addError(IPersistentVector error, int idx);

        int getIndex();

        Set<?> getErrors();
    }

    private static final IParseErrors NULL_PARSE_ERRORS = new IParseErrors() {
            public void addError(final IPersistentVector error, final int idx) {
            }

            public int getIndex () {
                return -1;
            }

            public Set<IPersistentVector> getErrors() {
                return null;
            }
        };

    public static final class ParseErrors implements IParseErrors {
        private final Set<IPersistentVector> errs;
        private int idx;

        public ParseErrors() {
            this.errs = new HashSet<>();
            this.idx = 0;
        }

        public void addError(final IPersistentVector error, final int idx) {
            if (this.idx == idx) {
                errs.add(error);
            } else if (this.idx < idx) {
                errs.clear();
                errs.add(error);
                this.idx = idx;
            }
        }

        public int getIndex() {
            return idx;
        }

        public Set<IPersistentVector> getErrors() {
            return errs;
        }
    }

    private static final IPersistentVector WS_ERROR = PersistentVector.create(Keyword.intern("expected"), "<WS>");

    private static int skipWhitespace(final Pattern pattern, final String in, final int idx, final IParseErrors errors) {
        final Matcher m = pattern.matcher(in).region(idx, in.length()).useTransparentBounds(true);
        if (m.lookingAt()) {
            return m.end();
        } else if (0 == idx) {
            return 0;
        } else {
            errors.addError(WS_ERROR, idx);
            return -1;
        }
    }

    public static abstract class AParser {
        public abstract ParseState parse(String in, int idx, ParseState[] memos, IParseErrors errors, final boolean hide);

        public AParser init(final AParser[] rules) {
            return this;
        }
    }

    public static final class EpsilonParser extends AParser {
        private static final IPersistentVector ERROR = PersistentVector.create(Keyword.intern("expected"), "<EOF>");
        private final Pattern wsPattern;

        public EpsilonParser(final Pattern wsPattern) {
            this.wsPattern = wsPattern;
        }

        public ParseState parse(final String in, int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            idx = skipWhitespace(wsPattern, in, idx, errors);
            if (idx != -1) {
                if (idx == in.length()) {
                    return new ParseState(PersistentVector.EMPTY, idx);
                } else {
                    errors.addError(ERROR, idx);
                    return null;
                }
            } else {
                return null;
            }
        }
    }

    public static final class NonTerminalParser extends AParser {
        private final int ruleId;

        public NonTerminalParser(final int ruleId) {
            this.ruleId = ruleId;
        }

        public ParseState parse(final String in, final int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            throw new UnsupportedOperationException();
        }

        public AParser init(final AParser[] rules) {
            return rules[ruleId];
        }
    }

    public static final Predicate<IPersistentVector> ALWAYS_RAW = new Predicate<IPersistentVector>() {
            public boolean test(final IPersistentVector ast) {
                return true;
            }
        };

    public static final Predicate<IPersistentVector> NEVER_RAW = new Predicate<IPersistentVector>() {
            public boolean test(final IPersistentVector ast) {
                return false;
            }
        };

    public static final Predicate<IPersistentVector> SINGLE_CHILD = new Predicate<IPersistentVector>() {
            public boolean test(final IPersistentVector ast) {
                return 1 == ast.count();
            }
        };

    public static final class RuleParser extends AParser {
        private static final Keyword START_IDX = Keyword.intern("start-idx");
        private static final Keyword END_IDX = Keyword.intern("end-idx");

        private final Keyword ruleName;
        private final Predicate<IPersistentVector> rawPred;
        private AParser parser;

        public RuleParser(final Keyword ruleName, Predicate<IPersistentVector> rawPred, final AParser parser) {
            this.ruleName = ruleName;
            this.rawPred = rawPred;
            this.parser = parser;
        }

        public ParseState parse(final String in, final int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            final ParseState state = parser.parse(in, idx, memos, errors, hide);
            if (state != null) {
                if (hide) {
                    return new ParseState(PersistentVector.EMPTY, state.idx);
                } else if (rawPred.test(state.ast)) {
                    return state;
                } else {
                    final IPersistentMap meta = new PersistentArrayMap(new Object[] {START_IDX, idx, END_IDX, state.idx});
                    ITransientCollection newAst = PersistentVector.EMPTY.asTransient();
                    newAst = newAst.conj(ruleName);
                    for (Object x : ((List<?>) state.ast)) {
                        newAst = newAst.conj(x);
                    }
                    return new ParseState(PersistentVector.create((Object) ((IObj) newAst.persistent()).withMeta(meta)), state.idx);
                }
            } else {
                return null;
            }
        }

        public AParser init(final AParser[] rules) {
            parser = parser.init(rules);
            return this;
        }
    }

    private static final ParseState NOT_FOUND = new ParseState(null, -1);

    private static final int MAX_RULE_ID = 512;
    public static final int RULE_ID_SHIFT = Integer.numberOfTrailingZeros(MAX_RULE_ID);

    public static final class MemoizeParser extends AParser {
        private final RuleParser parser;
        private final int ruleId;

        public MemoizeParser(final RuleParser parser, final int ruleId) {
            this.parser = parser;
            this.ruleId = ruleId;
        }

        public ParseState parse(final String in, final int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            final int memoIdx = ruleId | (idx << RULE_ID_SHIFT);
            ParseState state = memos[memoIdx];
            if (null == state) {
                state = parser.parse(in, idx, memos, errors, hide);
                memos[memoIdx] = state == null ? NOT_FOUND : state;
                return state;
            } else {
                return state == NOT_FOUND ? null : state;
            }
        }

        public AParser init(final AParser[] rules) {
            parser.init(rules);
            return this;
        }
    }

    public static final class MemoizeLeftRecParser extends AParser {
        private final RuleParser parser;
        private final int ruleId;

        public MemoizeLeftRecParser(final RuleParser parser, final int ruleId) {
            this.parser = parser;
            this.ruleId = ruleId;
        }

        public ParseState parse(final String in, final int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            final int memoIdx = ruleId | (idx << RULE_ID_SHIFT);
            ParseState state = memos[memoIdx];
            if (null == state) {
                state = NOT_FOUND;
                while (true) {
                    memos[memoIdx] = state;
                    final ParseState newState = parser.parse(in, idx, memos, errors, hide);
                    if (newState != null) {
                        if (state != NOT_FOUND && newState.idx <= state.idx) {
                            memos[memoIdx] = null;
                            return state;
                        } else {
                            state = newState;
                        }
                    } else {
                        memos[memoIdx] = null;
                        return state == NOT_FOUND ? null : state;
                    }
                }
            } else {
                return state == NOT_FOUND ? null : state;
            }
        }

        public AParser init(final AParser[] rules) {
            parser.init(rules);
            return this;
        }
    }

    public static final class HideParser extends AParser {
        private AParser parser;

        public HideParser(final AParser parser) {
            this.parser = parser;
        }

        public ParseState parse(final String in, final int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            return parser.parse(in, idx, memos, errors, true);
        }

        public AParser init(final AParser[] rules) {
            parser = parser.init(rules);
            return this;
        }
    }

    public static final class OptParser extends AParser {
        private AParser parser;

        public OptParser(final AParser parser) {
            this.parser = parser;
        }

        public ParseState parse(final String in, final int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            final ParseState state = parser.parse(in, idx, memos, errors, hide);
            if (state != null) {
                return state;
            } else {
                return new ParseState(PersistentVector.EMPTY, idx);
            }
        }

        public AParser init(final AParser[] rules) {
            parser = parser.init(rules);
            return this;
        }
    }

    public static final class NegParser extends AParser {
        private static final Keyword UNEXPECTED = Keyword.intern("unexpected");

        private AParser parser;

        public NegParser(final AParser parser) {
            this.parser = parser;
        }

        public ParseState parse(final String in, final int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            final ParseState state = parser.parse(in, idx, memos, NULL_PARSE_ERRORS, true);
            if (state != null) {
                errors.addError(PersistentVector.create(UNEXPECTED, in.substring(idx, state.idx)), idx);
                return null;
            } else {
                return new ParseState(PersistentVector.EMPTY, idx);
            }
        }

        public AParser init(final AParser[] rules) {
            parser = parser.init(rules);
            return this;
        }
    }

    public static final class RepeatParser extends AParser {
        private AParser parser;
        private final boolean isStar;

        public RepeatParser(final AParser parser, final boolean isStar) {
            this.parser = parser;
            this.isStar = isStar;
        }

        public ParseState parse(final String in, int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            final List<Object> ast = new ArrayList<>();
            boolean isMatch = false;
            while (true) {
                final ParseState state = parser.parse(in, idx, memos, errors, hide);
                if (state != null) {
                    isMatch = true;
                    idx = state.idx;
                    if (!hide) {
                        ast.addAll((List<?>) state.ast);
                    }
                } else {
                    if (isStar || isMatch) {
                        return new ParseState(PersistentVector.create(ast), idx);
                    } else {
                        return null;
                    }
                }
            }
        }

        public AParser init(final AParser[] rules) {
            parser = parser.init(rules);
            return this;
        }
    }

    public static final class CatParser extends AParser {
        private final AParser[] parsers;

        public CatParser(final List<AParser> parsers) {
            this.parsers = parsers.toArray(new AParser[parsers.size()]);
        }

        public ParseState parse(final String in, int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            final List<Object> ast = new ArrayList<>();
            for (int i = 0; i < parsers.length; i++) {
                final ParseState state = parsers[i].parse(in, idx, memos, errors, hide);
                if (state != null) {
                    idx = state.idx;
                    if (!hide) {
                        ast.addAll((List<?>) state.ast);
                    }
                } else {
                    return null;
                }
            }
            return new ParseState(PersistentVector.create(ast), idx);
        }

        public AParser init(final AParser[] rules) {
            for (int i = 0; i < parsers.length; i++) {
                parsers[i] = parsers[i].init(rules);
            }
            return this;
        }
    }

    public static final class AltParser extends AParser {
        private final AParser[] parsers;

        public AltParser(final List<AParser> parsers) {
            this.parsers = parsers.toArray(new AParser[parsers.size()]);
        }

        public ParseState parse(final String in, final int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            ParseState state1 = null;
            for (int i = 0; i < parsers.length; i++) {
                final ParseState state2 = parsers[i].parse(in, idx, memos, errors, hide);
                if (state1 == null || (state2 != null && state2.idx > state1.idx)) {
                    state1 = state2;
                }
            }
            return state1;
        }

        public AParser init(final AParser[] rules) {
            for (int i = 0; i < parsers.length; i++) {
                parsers[i] = parsers[i].init(rules);
            }
            return this;
        }
    }

    public static final class OrdParser extends AParser {
        private final AParser[] parsers;

        public OrdParser(final List<AParser> parsers) {
            this.parsers = parsers.toArray(new AParser[parsers.size()]);
        }

        public ParseState parse(final String in, final int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            for (int i = 0; i < parsers.length; i++) {
                final ParseState state = parsers[i].parse(in, idx, memos, errors, hide);
                if (state != null) {
                    return state;
                }
            }
            return null;
        }

        public AParser init(final AParser[] rules) {
            for (int i = 0; i < parsers.length; i++) {
                parsers[i] = parsers[i].init(rules);
            }
            return this;
        }
    }

    public static final class StringParser extends AParser {
        private final String string;
        private final IPersistentVector ast;
        private final IPersistentVector err;
        private final Pattern wsPattern;

        public StringParser(final String string, final Pattern wsPattern) {
            this.string = string;
            this.ast = PersistentVector.create(string);
            this.err = PersistentVector.create(Keyword.intern("expected"), string);
            this.wsPattern = wsPattern;
        }

        public ParseState parse(final String in, int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            idx = skipWhitespace(wsPattern, in, idx, errors);
            if (idx != -1) {
                if (in.regionMatches(true, idx, string, 0, string.length())) {
                    return new ParseState(hide ? PersistentVector.EMPTY : ast, idx + string.length());
                } else {
                    errors.addError(err, idx);
                    return null;
                }
            } else {
                return null;
            }
        }
    }

    public static final class RegexpParser extends AParser {
        private final Pattern pattern;
        private final IPersistentVector err;
        private final Function<Matcher, IPersistentVector> matcherFn;
        private final Pattern wsPattern;

        public RegexpParser(final Pattern pattern, final IPersistentVector err, final Function<Matcher, IPersistentVector> matcherFn, final Pattern wsPattern) {
            this.pattern = pattern;
            this.err = err;
            this.matcherFn = matcherFn;
            this.wsPattern = wsPattern;
        }

        public ParseState parse(final String in, int idx, final ParseState[] memos, final IParseErrors errors, final boolean hide) {
            idx = skipWhitespace(wsPattern, in, idx, errors);
            if (idx != -1) {
                final Matcher m = pattern.matcher(in).region(idx, in.length()).useTransparentBounds(true);
                if (m.lookingAt()) {
                    return new ParseState(hide ? PersistentVector.EMPTY : matcherFn.apply(m), m.end());
                } else {
                    errors.addError(err, idx);
                    return null;
                }
            } else {
                return null;
            }
        }
    }
}
